File: p1/p1_java/Carrera.java
import java.util.ArrayList;
import java.util.Random;

public abstract class Carrera {
    ArrayList<Bicicleta> bicicletas;
    String nombre;

    public Carrera(String s){
        bicicletas = new ArrayList<>();
        nombre = s;
    }

    public void inicio(){
        System.out.println("INICIO DE "+nombre);
        listarParticipantes();

        int total_bicis_a_retirar = getBicicletasRetiradas(); // Segun el tipo se retiran mas o menos
        
        Random indice_bici_a_eliminar = new Random();

        for(int i=0; i < total_bicis_a_retirar; i++){
            bicicletas.remove(indice_bici_a_eliminar.nextInt(bicicletas.size())); //del total de bicis, eliminamos una
        }
        System.out.println("***************\n");
        
    }

    public void fin(){
        System.out.println("FIN DE "+nombre);
        listarParticipantes();
        System.out.println("***************\n");
    }

    void aniadirBicicletas(ArrayList<Bicicleta> bicicletas){
        for (int i=0; i < bicicletas.size(); i++){
            this.bicicletas.add(bicicletas.get(i));
        }
    }

    abstract int getBicicletasRetiradas();

    private void listarParticipantes(){
        System.out.println(nombre+" --- Hay " + bicicletas.size()+ " participantes:");

        for (int i=0; i < bicicletas.size(); i++){
            System.out.println("Bici "+bicicletas.get(i).id);
        }
    }
}

File: p1/ejercicio2/Carrera.py
import random
#Definición de las Carreras
class Carrera():
    def __init__(self, nombre):
        self.nombre = nombre
        self.bicicletas = []

    def inicio(self):
        print("INICIO DE", self.nombre)
        print("***************\n")
        self.listar_participantes()
        total_bicis_a_retirar =self.get_bicicletas_retiradas()
        # for i in range(total_bicis_a_retirar):
        #     self.bicicletas.pop(random.Random.randint(0, len(self.bicicletas) - 1))  # Eliminamos una bicicleta del total
        indices_bicis_a_retirar = random.sample(range(len(self.bicicletas)), total_bicis_a_retirar)
        for index in sorted(indices_bicis_a_retirar, reverse=True):
            del self.bicicletas[index]
        print("***************\n")


    def fin(self):
        print("FIN DE", self.nombre)
        self.listar_participantes()
        print("***************\n")

    def aniadir_bicicletas(self, bicicletas):
        for bicicleta in bicicletas:
            self.bicicletas.append(bicicleta)

    def get_bicicletas_retiradas(self):
        pass


    def listar_participantes(self):
        print(self.nombre, "--- Hay", len(self.bicicletas), "participantes:")
        for bicicleta in self.bicicletas:
            print("Bici ", bicicleta.id)


class CarreraCarretera(Carrera):  
    def __init__(self, nombre):
        super().__init__(nombre)

    def get_bicicletas_retiradas(self):
        return int(len(self.bicicletas) * 0.1)

class CarreraMontana(Carrera):
    def __init__(self, nombre):
        super().__init__(nombre)

    def get_bicicletas_retiradas(self):
        return int(len(self.bicicletas) * 0.2)  
    
